Shrinking of Arbs (String, and maybe others?) exceeds min size limit

**Which version of Kotest are you using** 4.4.3


If I’ve got a String Arb and I set the minimum size to 1, it shrinks to size 0.

In this example, `authKey` shrinks to 0, and so gives an invalid reason for failure.

```kotlin

 test("expect 'actual code is failing', _not_ authKey length error") {
    checkAll(
      Arb.string(1..10, Arb.alphanumeric()),
    ) { authKey ->
      withClue("authKey must have length > 0") {
        authKey shouldHaveMinLength 1
      }
      fail("actual code is failing")
    }
  }
```

I’ve tried using this but then the test loops endlessly

```kotlin
Arb.string(1..10, Arb.alphanumeric()).filter { it.isEmpty() },
```

For fun I did a bit more hacking around. I think it produces some pretty nice output! I really like seeing the steps shrink.


```kotlin
  test("expect 'actual code is failing', _not_ authKey length error") {
    checkAll(
        Arb.authKey(1..100),
    ) { authKey ->
      withClue("authKey must have length > 0") {
        authKey shouldHaveMinLength 1
      }
      authKey shouldNotContain "&"
    }
  }
```

```
Property test failed for inputs

0) "8&\\7""#$DmSL#>p_TF4eq3"AFgEjc\Vig\rk`Eqr-_mEExokA;B8W"

Caused by java.lang.AssertionError: "8&\\7""#$DmSL#>p_TF4eq3"AFgEjc\Vig\rk`Eqr-_mEExokA;B8W" should not include substring "&" at (...)

Attempting to shrink arg "8&\\7""#$DmSL#>p_TF4eq3"AFgEjc\Vig\rk`Eqr-_mEExokA;B8W"
Shrink #1: "8&\\7""#$DmSL#>p_TF4eq3"AFg" fail
Shrink #2: "8&\\7""#$DmSL" fail
Shrink #3: "8&\\7"" fail
Shrink #4: "8&\" fail
Shrink #5: "8" pass
Shrink #6: "\" pass
Shrink #7: "&\" fail
Shrink #8: "&" fail
Shrink #9: "a" pass
Shrink result (after 9 shrinks) => "&"
```



```kotlin

fun Arb.Companion.authKey(
  range: IntRange = 1..100,
  codepoints: Arb<Codepoint> = Arb.ascii()
): Arb<String> = arbitrary(shrinker = StringShrinker2(range.first)) { rs: RandomSource ->
  Arb.string(range, codepoints).next(rs)
}

class StringShrinker2(
  private val minLength: Int = 0,
  private val simplestChar: Char = 'a',
) : Shrinker<String> {

  override fun shrink(value: String): List<String> {

    val isShortest = value.length == minLength
    val isSimplest = value.all { it == simplestChar }

    return when {
      isShortest && isSimplest -> emptyList()
      !isShortest              -> shorterVariants(value)
      !isSimplest              -> simplerVariants(value)
      else                     -> simplerVariants(value) + shorterVariants(value)
    }.map { it.padEnd(minLength, simplestChar) }
  }

  private fun simplerVariants(value: String) =
    listOf(
        value.replaceChar(simplestChar, CharSequence::indexOfFirst),
        value.replaceChar(simplestChar, CharSequence::indexOfLast),
        value.replaceChar(simplestChar) { value.getMidPoint() },
    )

  private fun shorterVariants(value: String) =
    listOf(
        value.take(value.getMidPoint()),
        value.takeLast(value.getMidPoint()),
        value.drop(1),
        value.dropLast(1),
    )

  private fun String.getMidPoint() =
    length.div(2).coerceIn(0..length)

  private fun String.replaceChar(
    newChar: Char,
    indexFinder: CharSequence.((Char) -> Boolean) -> Int,
  ): String {
    val index = indexFinder { it != newChar }.coerceIn(0..length)
    return replaceRange(index..index, newChar.toString())
  }
}
```

--- a/kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/strings.kt
+++ b/kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/strings.kt
@@ -96,34 +96,29 @@ class StringShrinkerWithMin(
 
       return when {
          isShortest && isSimplest -> emptyList()
-         !isShortest -> shorterVariants(value)
-         !isSimplest -> simplerVariants(value)
-         else -> (simplerVariants(value) + shorterVariants(value)).distinct()
-      }.map { it.padEnd(minLength, simplestChar) }
+         isShortest -> simplerVariants(value)
+         isSimplest -> shorterVariants(value)
+         else -> shorterVariants(value) + simplerVariants(value)
+      }.map { it.padEnd(minLength, simplestChar) }.distinct()
    }
 
    private fun simplerVariants(value: String) =
-      listOf(
-         value.replaceChar(simplestChar, value.indexOfFirst { it != simplestChar }),
-         value.replaceChar(simplestChar, value.indexOfLast { it != simplestChar }),
-         value.replaceChar(simplestChar, value.getMidPoint()),
-      )
+      listOfNotNull(replaceFirst(value, simplestChar), replaceLast(value, simplestChar))
 
    private fun shorterVariants(value: String) =
       listOf(
-         value.take(value.getMidPoint()),
-         value.takeLast(value.getMidPoint()),
+         value.take(value.length / 2 + value.length % 2),
+         value.takeLast(value.length / 2),
          value.drop(1),
          value.dropLast(1),
       )
 
-   private fun String.getMidPoint() =
-      length.div(2).coerceIn(0..length)
+   private fun replaceFirst(value: String, newChar: Char): String? =
+      replace(value, newChar, value.indexOfFirst { it != simplestChar })
 
-   private fun String.replaceChar(
-      newChar: Char,
-      index: Int,
-   ): String {
-      return replaceRange(index..index, newChar.toString())
-   }
+   private fun replaceLast(value: String, newChar: Char): String? =
+      replace(value, newChar, value.indexOfLast { it != simplestChar })
+
+   private fun replace(value: String, newChar: Char, index: Int) =
+      if (index == -1) null else value.replaceRange(index..index, newChar.toString())
 }

--- a/kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/strings.kt
+++ b/kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/strings.kt
@@ -29,7 +29,7 @@ fun Arb.Companion.string(
    else
       listOfNotNull(min, minPlus1).filter { it.length in minSize..maxSize }
 
-   return arbitrary(edgecases, StringShrinker) { rs ->
+   return arbitrary(edgecases, StringShrinkerWithMin(minSize)) { rs ->
       val size = rs.random.nextInt(minSize..maxSize)
       codepoints.take(size, rs).joinToString("") { it.asString() }
    }
@@ -56,6 +56,7 @@ fun Arb.Companion.string(range: IntRange, codepoints: Arb<Codepoint> = Arb.ascii
 fun Arb.Companion.string(size: Int, codepoints: Arb<Codepoint> = Arb.ascii()): Arb<String> =
    Arb.string(size, size, codepoints)
 
+@Deprecated("use StringShrinkerWithMin. This will be removed in 4.7")
 object StringShrinker : Shrinker<String> {
 
    override fun shrink(value: String): List<String> {
@@ -80,5 +81,49 @@ object StringShrinker : Shrinker<String> {
             )
          }
       }
+   }
+}
+
+class StringShrinkerWithMin(
+   private val minLength: Int = 0,
+   private val simplestChar: Char = 'a',
+) : Shrinker<String> {
+
+   override fun shrink(value: String): List<String> {
+
+      val isShortest = value.length == minLength
+      val isSimplest = value.all { it == simplestChar }
+
+      return when {
+         isShortest && isSimplest -> emptyList()
+         !isShortest -> shorterVariants(value)
+         !isSimplest -> simplerVariants(value)
+         else -> (simplerVariants(value) + shorterVariants(value)).distinct()
+      }.map { it.padEnd(minLength, simplestChar) }
+   }
+
+   private fun simplerVariants(value: String) =
+      listOf(
+         value.replaceChar(simplestChar, value.indexOfFirst { it != simplestChar }),
+         value.replaceChar(simplestChar, value.indexOfLast { it != simplestChar }),
+         value.replaceChar(simplestChar, value.getMidPoint()),
+      )
+
+   private fun shorterVariants(value: String) =
+      listOf(
+         value.take(value.getMidPoint()),
+         value.takeLast(value.getMidPoint()),
+         value.drop(1),
+         value.dropLast(1),
+      )
+
+   private fun String.getMidPoint() =
+      length.div(2).coerceIn(0..length)
+
+   private fun String.replaceChar(
+      newChar: Char,
+      index: Int,
+   ): String {
+      return replaceRange(index..index, newChar.toString())
    }
 }

--- a/kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/ShrinkingTest.kt
+++ b/kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/ShrinkingTest.kt
@@ -19,18 +19,46 @@ class ShrinkingTest : FunSpec() {
                }
             }
          }
-         println(stdout)
 
-         stdout.shouldContain("Property test failed for inputs")
-         stdout.shouldContain("0) \"\"aaVTT\$H<<3&rYlvW==(>]8Sga\"")
-         stdout.shouldContain("1) \"Z7XHs_\\&5YUAua4^Eq>m\$w:-\\EtY\\3b<\$<[WnDk,e'T9I3?1+Rvx;\")63cwl!`\"")
-         stdout.shouldContain("Caused by java.lang.AssertionError: 88 should be < 4 at")
-         stdout.shouldContain("Caused by java.lang.AssertionError: 62 should be < 4 at")
-         stdout.shouldContain("Caused by java.lang.AssertionError: 4 should be < 4 at")
-         stdout.shouldContain("Attempting to shrink arg \"\"aaVTT\$H<<3&rYlvW==(>]8Sga\"")
-         stdout.shouldContain("Attempting to shrink arg \"Z7XHs_\\&5YUAua4^Eq>m\$w:-\\EtY\\3b<\$<[WnDk,e'T9I3?1+Rvx;\")63cwl!`\"")
-         stdout.shouldContain("Shrink result (after 6 shrinks) => <empty string>")
-         stdout.shouldContain("Shrink result (after 10 shrinks) => \"aaaa\"")
+         stdout.shouldContain(
+            """
+Attempting to shrink arg ""aaVTT${"$"}H<<3&rYlvW==(>]8Sga"
+Shrink #1: ""aaVTT${"$"}H<<3&r" fail
+Shrink #2: ""aaVTT${'$'}" fail
+Shrink #3: ""aaV" fail
+Shrink #4: ""a" fail
+Shrink #5: ""${'"'} fail
+Shrink #6: <empty string> fail
+Shrink result (after 6 shrinks) => <empty string>
+            """.trim()
+         )
+
+         stdout.shouldContain(
+            """
+Attempting to shrink arg "Z7XHs_\&5YUAua4^Eq>m${"$"}w:-\EtY\3b<${'$'}<[WnDk,e'T9I3?1+Rvx;")63cwl!`"
+Shrink #1: "Z7XHs_\&5YUAua4^Eq>m${"$"}w:-\EtY\3b" fail
+Shrink #2: "Z7XHs_\&5YUAua4^" fail
+Shrink #3: "Z7XHs_\&" fail
+Shrink #4: "Z7XH" fail
+Shrink #5: "Z7" pass
+Shrink #6: "XH" pass
+Shrink #7: "7XH" pass
+Shrink #8: "Z7X" pass
+Shrink #9: "a7XH" fail
+Shrink #10: "a7" pass
+Shrink #11: "a7X" pass
+Shrink #12: "aaXH" fail
+Shrink #13: "aa" pass
+Shrink #14: "aXH" pass
+Shrink #15: "aaX" pass
+Shrink #16: "aaaH" fail
+Shrink #17: "aH" pass
+Shrink #18: "aaH" pass
+Shrink #19: "aaa" pass
+Shrink #20: "aaaa" fail
+Shrink result (after 20 shrinks) => "aaaa"
+            """.trim()
+         )
       }
    }
 }

--- a/kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/StringShrinkerTest.kt
+++ b/kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/StringShrinkerTest.kt
@@ -115,10 +115,11 @@ class StringShrinkerTest : DescribeSpec({
          stdout.shouldContain(
             """
 Attempting to shrink arg ")E2v${'$'}"
-Shrink #1: ")Eaa" fail
-Shrink #2: "aEaa" fail
-Shrink #3: "aaaa" fail
-Shrink result (after 3 shrinks) => "aaaa"
+Shrink #1: ")E2a" fail
+Shrink #2: "aE2a" fail
+Shrink #3: "aa2a" fail
+Shrink #4: "aaaa" fail
+Shrink result (after 4 shrinks) => "aaaa"
             """.trim()
          )
          PropertyTesting.shouldPrintShrinkSteps = prt

--- a/kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/StringShrinkerTest.kt
+++ b/kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/StringShrinkerTest.kt
@@ -1,21 +1,28 @@
 package com.sksamuel.kotest.property.shrinking
 
+import io.kotest.assertions.shouldFail
+import io.kotest.core.spec.style.DescribeSpec
 import io.kotest.core.spec.style.StringSpec
+import io.kotest.extensions.system.captureStandardOut
 import io.kotest.inspectors.forAtLeastOne
 import io.kotest.matchers.comparables.shouldBeLessThan
 import io.kotest.matchers.shouldBe
+import io.kotest.matchers.string.shouldContain
 import io.kotest.matchers.string.shouldEndWith
 import io.kotest.matchers.string.shouldHaveLength
 import io.kotest.matchers.string.shouldNotContain
 import io.kotest.matchers.string.shouldStartWith
+import io.kotest.property.Arb
+import io.kotest.property.PropTestConfig
 import io.kotest.property.PropertyTesting
 import io.kotest.property.ShrinkingMode
 import io.kotest.property.arbitrary.StringShrinker
+import io.kotest.property.arbitrary.string
 import io.kotest.property.checkAll
 import io.kotest.property.internal.doShrinking
 import io.kotest.property.rtree
 
-class StringShrinkerTest : StringSpec({
+class StringShrinkerTest : DescribeSpec({
 
    beforeSpec {
       PropertyTesting.shouldPrintShrinkSteps = false
@@ -25,69 +32,96 @@ class StringShrinkerTest : StringSpec({
       PropertyTesting.shouldPrintShrinkSteps = true
    }
 
-   "StringShrinker should include bisected input" {
-      checkAll { a: String ->
-         if (a.length > 1) {
-            val candidates = StringShrinker.shrink(a)
-            candidates.forAtLeastOne {
-               it.shouldHaveLength(a.length / 2 + a.length % 2)
-            }
-            candidates.forAtLeastOne {
-               it.shouldHaveLength(a.length / 2)
-            }
-         }
-      }
-   }
+   describe("StringShrinker should") {
+      it("include bisected input") {
+         checkAll { a: String ->
+            if (a.length > 1) {
+               val candidates = StringShrinker.shrink(a)
+               candidates.forAtLeastOne {
+                  it.shouldHaveLength(a.length / 2 + a.length % 2)
+               }
+               candidates.forAtLeastOne {
+                  it.shouldHaveLength(a.length / 2)
+               }
 
-   "StringShrinker should include 2 padded 'a's" {
-      checkAll { a: String ->
-         if (a.length > 1) {
-            val candidates = StringShrinker.shrink(a)
-            candidates.forAtLeastOne {
-               it.shouldEndWith("a".repeat(a.length / 2))
-            }
-            candidates.forAtLeastOne {
-               it.shouldStartWith("a".repeat(a.length / 2))
             }
          }
       }
-   }
 
-   "StringShrinker should shrink to expected value" {
-      val prt = PropertyTesting.shouldPrintShrinkSteps
-      PropertyTesting.shouldPrintShrinkSteps = false
+      it("should include 2 padded 'a's") {
+         checkAll { a: String ->
+            if (a.length > 1) {
+               val candidates = StringShrinker.shrink(a)
+               candidates.forAtLeastOne {
+                  it.shouldEndWith("a".repeat(a.length / 2))
+               }
+               candidates.forAtLeastOne {
+                  it.shouldStartWith("a".repeat(a.length / 2))
+               }
+            }
+         }
+      }
 
-      checkAll<String> { a ->
+      it("shrink to expected value") {
+         val prt = PropertyTesting.shouldPrintShrinkSteps
+         PropertyTesting.shouldPrintShrinkSteps = false
 
+         checkAll<String> { a ->
+
+            val shrinks = StringShrinker.rtree(a)
+            val shrunk = doShrinking(shrinks, ShrinkingMode.Unbounded) {
+               it.shouldNotContain("#")
+            }
+
+            if (a.contains("#")) {
+               shrunk.shrink shouldBe "#"
+            } else {
+               shrunk.shrink shouldBe a
+            }
+         }
+
+         PropertyTesting.shouldPrintShrinkSteps = prt
+      }
+
+      it("prefer padded values") {
+         val prt = PropertyTesting.shouldPrintShrinkSteps
+         PropertyTesting.shouldPrintShrinkSteps = false
+
+         val a = "97asd!@#ASD'''234)*safmasd"
          val shrinks = StringShrinker.rtree(a)
-         val shrunk = doShrinking(shrinks, ShrinkingMode.Unbounded) {
-            it.shouldNotContain("#")
-         }
+         doShrinking(shrinks, ShrinkingMode.Unbounded) {
+            it.length.shouldBeLessThan(3)
+         }.shrink shouldBe "aaa"
 
-         if (a.contains("#")) {
-            shrunk.shrink shouldBe "#"
-         } else {
-            shrunk.shrink shouldBe a
-         }
+         doShrinking(shrinks, ShrinkingMode.Unbounded) {
+            it.length.shouldBeLessThan(8)
+         }.shrink shouldBe "aaaaaaaa"
+
+         PropertyTesting.shouldPrintShrinkSteps = prt
       }
 
-      PropertyTesting.shouldPrintShrinkSteps = prt
-   }
-
-   "StringShrinker should prefer padded values" {
-      val prt = PropertyTesting.shouldPrintShrinkSteps
-      PropertyTesting.shouldPrintShrinkSteps = false
-
-      val a = "97asd!@#ASD'''234)*safmasd"
-      val shrinks = StringShrinker.rtree(a)
-      doShrinking(shrinks, ShrinkingMode.Unbounded) {
-         it.length.shouldBeLessThan(3)
-      }.shrink shouldBe "aaa"
-
-      doShrinking(shrinks, ShrinkingMode.Unbounded) {
-         it.length.shouldBeLessThan(8)
-      }.shrink shouldBe "aaaaaaaa"
-
-      PropertyTesting.shouldPrintShrinkSteps = prt
+      it("respect min value") {
+         val prt = PropertyTesting.shouldPrintShrinkSteps
+         PropertyTesting.shouldPrintShrinkSteps = true
+         val stdout = captureStandardOut {
+            shouldFail {
+               checkAll(PropTestConfig(seed = 123123), Arb.string(4, 8)) { a ->
+                  // will cause the value to fail and shrinks be used, but nothing should be shrunk
+                  // past the min value of 4, even though we fail on anything >= 2
+                  a.shouldHaveLength(1)
+               }
+            }
+         }
+         stdout.shouldContain(
+            """
+Attempting to shrink arg ")E2v${'$'}"
+Shrink #1: ")Eaa" fail
+Shrink #2: "aEaa" fail
+Shrink #3: "aaaa" fail
+Shrink result (after 3 shrinks) => "aaaa"
+            """.trim()
+         )
+         PropertyTesting.shouldPrintShrinkSteps = prt
+      }
    }
 })

